src/graph/CreateSpaceExecutor.cpp

```diff
@@ -17,36 +19,110 @@ CreateSpaceExecutor::CreateSpaceExecutor(Sentence *sentence,
                 }
                 break;
-            case SpaceOptItem::REPLICA_FACTOR:
-                replicaFactor_ = item->get_replica_factor();
-                if (replicaFactor_ <= 0) {
+            }
+            case SpaceOptItem::REPLICA_FACTOR: {
+                spaceDesc_.replicaFactor_ = item->get_replica_factor();
+                if (spaceDesc_.replicaFactor_ <= 0) {
                     return Status::Error("Replica_factor value should be greater than zero");
                 }
                 break;
+            }
+            case SpaceOptItem::CHARSET: {
+                result = item->get_charset();
+                folly::toLowerAscii(result);
+                retStatus = charsetInfo->isSupportCharset(result);
+                if (!retStatus.ok()) {
+                    return retStatus;
+                }
+                spaceDesc_.charsetName_ = std::move(result);
+                break;
+            }
+            case SpaceOptItem::COLLATE: {
+                result = item->get_collate();
+                folly::toLowerAscii(result);
+                retStatus = charsetInfo->isSupportCollate(result);
+                if (!retStatus.ok()) {
+                    return retStatus;
+                }
+                spaceDesc_.collationName_ = std::move(result);
+                break;
+            }
+        }
+    }
+
+    // if charset and collate are not specified, set default value
+    if (!spaceDesc_.charsetName_.empty() && !spaceDesc_.collationName_.empty()) {
+        retStatus = charsetInfo->charsetAndCollateMatch(spaceDesc_.charsetName_,
+                                                        spaceDesc_.collationName_);
+        if (!retStatus.ok()) {
+            return retStatus;
+        }
+    } else if (!spaceDesc_.charsetName_.empty()) {
+        retStatusOr = charsetInfo->getDefaultCollationbyCharset(spaceDesc_.charsetName_);
+        if (!retStatusOr.ok()) {
+            return retStatusOr.status();
+        }
+        spaceDesc_.collationName_ = std::move(retStatusOr.value());
+    } else if (!spaceDesc_.collationName_.empty()) {
+        retStatusOr = charsetInfo->getCharsetbyCollation(spaceDesc_.collationName_);
+        if (!retStatusOr.ok()) {
+            return retStatusOr.status();
         }
+        spaceDesc_.charsetName_ = std::move(retStatusOr.value());
     }
+
+    if (spaceDesc_.charsetName_.empty() && spaceDesc_.collationName_.empty()) {
+        std::string charsetName = FLAGS_default_charset;
+        folly::toLowerAscii(charsetName);
+        retStatus = charsetInfo->isSupportCharset(charsetName);
+        if (!retStatus.ok()) {
+            return retStatus;
+        }
+
+        std::string collateName = FLAGS_default_collate;
+        folly::toLowerAscii(collateName);
+        retStatus = charsetInfo->isSupportCollate(collateName);
+        if (!retStatus.ok()) {
+            return retStatus;
+        }
+
```

Comment:
A question,  could we simplify code as below?
```
if (spaceDesc_.charsetName_.empty()) {
} ;
if (spaceDesc_.collationName_.empty()) {
};
retStatus = charsetInfo->charsetAndCollateMatch(spaceDesc_.charsetName_, spaceDesc_.collationName_);
```

Reply:
hen only one charsetName or collateName is empty, get the default value at this time. At this time, you don't need to check whether charsetName and collateName match, because when you get the other value, the default is matched.

When both charsetName and collateName are empty, this uses the value of the configuration file or the default value of FLAGs. At this time, you need to check whether it matches. In case someone has intentionally written the configuration file incorrectly

Reply:
Excellent. got.